### PR TITLE
feat(menu): Status Badge and Menu and menu badge, Checkable menus

### PR DESCRIPTION
**Pull request by SiTaggart**
Description and useful links to discussions / issues / tickets

### Status Badge

Create a composed component of Badge that follows out status pattern and creates a component that ties color to icon with specific status variants. It's a seperate component as I didn't want to overload base badge, but it is exported from the package.

### Default badge update

Change to be white for use in TopBar

### Menu Badge

Ability to use badge as a part of a menu trigger. The menu button is a chevron contained in the badge. The Menu badge can take any badge variant

### Status Menu

Ability to use the Status badge along with new Status based menu items to create status menus. Composed examples that pair color and icons again to make it easier to do the right thing.

### Checkable menu items

Creates MenuItemCheckbox and MenuItemRadio to allow users to check menu items. Can be useful to view selection, Rich Text Editor toolbars, account switching

## Checklist

- [ ] Work in a "Draft" branch whilst you are getting feedback to reduce Chromatic costs. Once you are ready for approval, convert the PR to "Ready to review"
- [ ] Ensure all `required` checks have passed
- [ ] Icon changes must have a product design review
- [ ] Website content changes must have a product design review
- [ ] At least two engineers have reviewed any code changes
- [ ] At least one engineering lead has reviewed any core package changes or repository infrastructure
- [ ] Website visual regression testing is run by adding `🕵🏻‍♀️ Run website visual regression` label

### CODE REVIEW

File: packages/paste-core/components/menu/__tests__/index.spec.tsx [hunk 1]
```diff
@@ -53,6 +62,18 @@ const MenuMock: React.FC<React.PropsWithChildren<{groupRef?: React.Ref<HTMLDivEl
             Search with Bing
           </MenuItem>
         </MenuGroup>
+        <MenuItemCheckbox {...menu} name="display-view" value="wrap">
+          Check Wrap
+        </MenuItemCheckbox>
+        <MenuItemCheckbox {...menu} name="display-view" value="no-wrap">
+          Check No Wrap
+        </MenuItemCheckbox>
+        <MenuItemRadio {...menu} name="display-view" value="wrap">
```
Comment (TheSisb) [inline packages/paste-core/components/menu/__tests__/index.spec.tsx:71]: suggestion: the radios should use a different name than the checkboxes

File: packages/paste-core/components/menu/src/MenuBadge.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        </MenuBadgeContent>
+      </Badge>
+    );
+  }
+);
+
+MenuBadge.displayName = 'MenuBadge';
+
+MenuBadge.propTypes = {
+  children: PropTypes.node.isRequired,
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuBadge.tsx:55]: todo: missing props

File: packages/paste-core/components/menu/src/MenuBadge.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        {children}
+        <MenuPrimitiveButton
+          {...props}
+          variant="secondary_icon"
+          size="reset"
+          element={`${element}_BUTTON`}
+          as={Button}
+          ref={ref}
+        >
+          <ChevronDownIcon decorative={false} size="sizeIcon10" title={buttonLabel} />
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuBadge.tsx:32]: question: title or `<ScreenReaderOnly>` like we do elsewhere? (ModalHeader, DataGridHeaderSort, etc)

File: packages/paste-core/components/menu/src/MenuBadge.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    );
+  }
+);
+MenuBadgeContent.displayName = 'MenuBadgeContent';
+
+const MenuBadge = React.forwardRef<HTMLButtonElement, MenuBadgeProps>(
+  ({children, element = 'MENU_BADGE', variant = 'default', buttonLabel, ...props}, ref) => {
+    return (
+      <Badge element={element} as="span" variant={variant}>
+        <MenuBadgeContent {...props} element={element} buttonLabel={buttonLabel} ref={ref}>
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuBadge.tsx:44]: suggestion: prefix the label with i18n, so `i18nButtonLabel`. So it matches what we do in Modals for the Close button and works out of the box with Consoles i18n linter (I think)

File: packages/paste-core/components/menu/src/MenuBadgeStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        </MenuBadgeContent>
+      </StatusBadge>
+    );
+  }
+);
+
+MenuBadgeStatus.displayName = 'MenuBadgeStatus';
+
+MenuBadgeStatus.propTypes = {
+  children: PropTypes.node.isRequired,
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuBadgeStatus.tsx:28]: todo: missing proptypes

File: packages/paste-core/components/menu/src/MenuItem.tsx [hunk 6]
```diff
@@ -105,3 +113,29 @@ const MenuItem = React.forwardRef<HTMLDivElement, MenuItemProps>(
 );
 MenuItem.displayName = 'MenuItem';
 export {MenuItem};
+
+const MenuItemCheckbox = React.forwardRef<HTMLDivElement, MenuItemCheckboxProps>(
+  ({as = StyledMenuItem, variant: _variant, element = 'MENU_ITEM_CHECKBOX', ...props}, ref) => {
+    let variant = _variant;
+    const isGrouped = React.useContext(MenuGroupContext) === MenuItemVariants.GROUP_ITEM;
+    if (isGrouped && _variant === MenuItemVariants.DESTRUCTIVE) variant = MenuItemVariants.DESTRUCTIVE_GROUP_ITEM;
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuItem.tsx:121]: suggestion: prefer a nested if for readability
```
if (isGrouped) {
  if (variant === MenuItemVariants.DESTRUCTIVE) {
    variant = MenuItemVariants.DESTRUCTIVE_GROUP_ITEM;
  } else {
    variant = MenuItemVariants.GROUP_ITEM;
  }
}
```

File: packages/paste-core/components/menu/stories/index.stories.tsx [hunk 7]
```diff
@@ -271,6 +281,8 @@ export const MenuDropdown = (): React.ReactNode => {
     registerItem: () => {},
     unregisterItem: () => {},
     tabIndex: -1,
+    unstable_values: {},
```
Comment (TheSisb) [inline packages/paste-core/components/menu/stories/index.stories.tsx:284]: question: why add these?

Comment (SiTaggart) [reply packages/paste-core/components/menu/stories/index.stories.tsx:284]: Typescript made me do it for checkbox and radio menu items

File: packages/paste-core/components/menu/src/MenuBadge.tsx (hunk 4 above)
Comment (SiTaggart) [reply packages/paste-core/components/menu/src/MenuBadge.tsx:44]: Probably best to follow the rest of the system

File: packages/paste-core/components/menu/__tests__/menu-badge-customization.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      placement: 'bottom-start',
+      previous: NOOP,
+      registerGroup: NOOP,
+      registerItem: NOOP,
+      reset: NOOP,
+      rtl: false,
+      setAnimated: NOOP,
+      setBaseId: NOOP,
+      setCurrentId: NOOP,
+      setLoop: NOOP,
```
Comment (TheSisb) [inline packages/paste-core/components/menu/__tests__/menu-badge-customization.spec.tsx:100]: 0_0

File: packages/paste-core/components/menu/stories/menu-checkable-items.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  MenuButton,
+  MenuItemCheckbox,
+  SubMenuButton,
+  MenuGroup,
+  MenuItem,
+  MenuSeparator,
+  useMenuState,
+  MenuItemRadio,
+} from '../src';
+import {Box} from '@twilio-paste/box';
```
Comment (TheSisb) [inline packages/paste-core/components/menu/stories/menu-checkable-items.stories.tsx:25]: praise: github warnings coming in clutch

File: packages/paste-core/components/menu/__tests__/menu-badge-customization.spec.tsx (hunk 8 above)
Comment (SiTaggart) [reply packages/paste-core/components/menu/__tests__/menu-badge-customization.spec.tsx:100]: It's pure copy and paste from existing tests from Menu

**Review by loreina — DISMISSED** [review]

**Review by loreina — DISMISSED** [review]

File: packages/paste-core/components/badge/src/Badge.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        {variant === 'error_counter' ? <ErrorIcon element={`${element}_ICON`} decorative size="sizeIcon10" /> : null}
+        {resizedChildren}
+      </Box>
+    );
+  }
+);
+
+Badge.displayName = 'Badge';
+
+Badge.propTypes = {...BadgePropTypes};
```
Comment (TheSisb) [inline packages/paste-core/components/badge/src/Badge.tsx:46]:
```suggestion
Badge.propTypes = BadgePropTypes;
```

File: packages/paste-core/components/badge/src/StatusBadge.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
```
Comment (TheSisb) [inline packages/paste-core/components/badge/src/StatusBadge.tsx:1]: Normally package exports should be prefixed with the package name, so something like `BadgeStatus`. I like `StatusBadge` more, but we also have a similar thing in Menu in this PR: `MenuBadgeStatus`.

So either we rename this `BadgeStatus` 🤢  or we renamed `MenuBadgeStatus` to `MenuStatusBadge`

File: packages/paste-core/components/menu/src/MenuButton.tsx [hunk 12]
```diff
@@ -14,4 +14,7 @@ const MenuButton = React.forwardRef<HTMLButtonElement, MenuButtonProps>(({elemen
   );
 });
 MenuButton.displayName = 'MenuButton';
+MenuButton.propTypes = {
+  ...ButtonPropTypes,
+};
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuButton.tsx:19]:
```suggestion
MenuButton.propTypes = ButtonPropTypes;
```

File: packages/paste-core/components/menu/src/MenuItem.tsx [hunk 13]
```diff
@@ -85,23 +107,90 @@ export const StyledMenuItem = React.forwardRef<HTMLDivElement | HTMLAnchorElemen
...
+  id: PropTypes.string,
+  onClick: PropTypes.func,
+  element: PropTypes.string,
+  value: PropTypes.string,
+  checked: PropTypes.bool,
+  name: PropTypes.string,
+};
+export {MenuItemCheckbox};
+
+const MenuItemRadio = React.forwardRef<HTMLDivElement, MenuItemRadioProps>(
```
Comment (TheSisb) [inline packages/paste-core/components/menu/src/MenuItem.tsx:176]: suggestion (not-blocking): kinda feel like some of the things in this file can be cleanly pulled out into their own files

Comment (SiTaggart) [reply packages/paste-core/components/menu/src/MenuItem.tsx:176]: File per component probably, it is getting a little messy

File: packages/paste-core/components/badge/src/StatusBadge.tsx (hunk 11 above)
Comment (SiTaggart) [reply packages/paste-core/components/badge/src/StatusBadge.tsx:1]: Yeah, it's awkward. I too really dislike `BadgeStatus`. The problem I faced with Menu was that there is actually a `MenuBadge`, and I was really struggling with:

- `MenuButton`
- `MenuBadge`
- `MenuStatusBadge`

Gunna throw a straw poll up in Slack to find our least worst preference

**Pull request merged**